Lib/test/test_capi/test_number.py

```diff
@@ -0,0 +1,349 @@
+import itertools
+import operator
+import unittest
+import sys
+import warnings
+
+from test.support import cpython_only, import_helper
+from _testbuffer import ndarray
+
+_testcapi = import_helper.import_module('_testcapi')
+from _testcapi import PY_SSIZE_T_MIN, PY_SSIZE_T_MAX
+
+NULL = None
+
+class BadDescr:
+    def __get__(self, obj, objtype=None):
+        raise RuntimeError
+
+class WithDunder:
+    def _meth(self, *args):
+        if self.val:
+            return self.val
+        if self.exc:
+            raise self.exc
+    @classmethod
+    def with_val(cls, val):
+        obj = super().__new__(cls)
+        obj.val = val
+        obj.exc = None
+        setattr(cls, cls.methname, cls._meth)
+        return obj
+
+    @classmethod
+    def with_exc(cls, exc):
+        obj = super().__new__(cls)
+        obj.val = None
+        obj.exc = exc
+        setattr(cls, cls.methname, cls._meth)
+        return obj
+
+class HasBadAttr:
+    def __new__(cls):
+        obj = super().__new__(cls)
+        setattr(cls, cls.methname, BadDescr())
+        return obj
+
+
+class IndexLike(WithDunder):
+    methname = '__index__'
+
+class IntLike(WithDunder):
+    methname = '__int__'
+
+class FloatLike(WithDunder):
+    methname = '__float__'
+
+class HasTrunc(WithDunder):
+    methname = '__trunc__'
+
+
+class HasBadTrunc:
+    __trunc__ = BadDescr()
+
+
+def subclassof(base):
+    return type(base.__name__ + 'Subclass', (base,), {})
+
+
+class SomeError(Exception):
+    pass
+
+class OtherError(Exception):
+    pass
+
+
+class CAPITest(unittest.TestCase):
+    def test_check(self):
+        # Test PyNumber_Check()
+        check = _testcapi.number_check
+
+        self.assertTrue(check(1))
+        self.assertTrue(check(IndexLike.with_val(1)))
+        self.assertTrue(check(IntLike.with_val(99)))
+        self.assertTrue(check(0.5))
+        self.assertTrue(check(FloatLike.with_val(4.25)))
+        self.assertTrue(check(1+2j))
+
+        self.assertFalse(check([]))
+        self.assertFalse(check("1 + 1j"))
```

Comment:
This test is misleading :-)

```suggestion
        self.assertFalse(check("abc"))
```